File: backend/src/models/Task.ts
import { db } from '../services/db.service';
import { Task, TaskStatus, TaskPriority } from '../types';
import { v4 as uuidv4 } from 'uuid';

export class TaskModel {
    /**
     * Get all tasks with optional filters
     */
    static async findAll(filters?: {
        agentId?: string;
        status?: TaskStatus;
        userId?: string;
        limit?: number;
        offset?: number;
    }): Promise<{ tasks: Task[], total: number }> {
        let query = 'SELECT * FROM tasks WHERE 1=1';
        let countQuery = 'SELECT COUNT(*)::int as total FROM tasks WHERE 1=1';
        const params: any[] = [];
        let paramCount = 1;

        if (filters?.agentId) {
            const clause = ` AND agent_id = $${paramCount++}`;
            query += clause;
            countQuery += clause;
            params.push(filters.agentId);
        }
        if (filters?.status) {
            const clause = ` AND status = $${paramCount++}`;
            query += clause;
            countQuery += clause;
            params.push(filters.status);
        }
        if (filters?.userId) {
            const clause = ` AND user_id = $${paramCount++}`;
            query += clause;
            countQuery += clause;
            params.push(filters.userId);
        }

        query += ' ORDER BY created_at DESC';

        if (filters?.limit) {
            query += ` LIMIT $${paramCount++}`;
            params.push(filters.limit);
        }
        if (filters?.offset) {
            query += ` OFFSET $${paramCount++}`;
            params.push(filters.offset);
        }

        // Execute both queries
        // Note: For count query, we only need the filter params, not limit/offset
        const filterParams = params.slice(0, paramCount - 1 - (filters?.limit ? 1 : 0) - (filters?.offset ? 1 : 0));

        const [result, countResult] = await Promise.all([
            db.query(query, params),
            db.query(countQuery, filterParams)
        ]);

        return {
            tasks: result.rows,
            total: countResult.rows[0].total
        };
    }

    /**
     * Get task by ID
     */
    static async findById(id: string): Promise<Task | null> {
        const result = await db.query('SELECT * FROM tasks WHERE id = $1', [id]);
        return result.rows[0] || null;
    }

    /**
     * Create new task
     */
    static async create(data: Omit<Task, 'id' | 'createdAt'>): Promise<Task> {
        const id = uuidv4();
        const result = await db.query(
            `INSERT INTO tasks (id, agent_id, user_id, action, input, status, priority)
       VALUES ($1, $2, $3, $4, $5, $6, $7)
       RETURNING *`,
            [
                id,
                data.agentId,
                data.userId || null,
                data.action,
                JSON.stringify(data.input),
                data.status,
                data.priority,
            ]
        );
        return result.rows[0];
    }

    /**
     * Update task
     */
    static async update(id: string, data: Partial<Task>): Promise<Task | null> {
        const updates: string[] = [];
        const values: any[] = [];
        let paramCount = 1;

        if (data.status !== undefined) {
            updates.push(`status = $${paramCount++}`);
            values.push(data.status);
        }
        if (data.output !== undefined) {
            updates.push(`output = $${paramCount++}`);
            values.push(JSON.stringify(data.output));
        }
        if (data.error !== undefined) {
            updates.push(`error = $${paramCount++}`);
            values.push(data.error);
        }
        if (data.startedAt !== undefined) {
            updates.push(`started_at = $${paramCount++}`);
            values.push(data.startedAt);
        }
        if (data.completedAt !== undefined) {
            updates.push(`completed_at = $${paramCount++}`);
            values.push(data.completedAt);
        }

        if (updates.length === 0) {
            return await this.findById(id);
        }

        values.push(id);

        const result = await db.query(
            `UPDATE tasks SET ${updates.join(', ')} WHERE id = $${paramCount} RETURNING *`,
            values
        );

        return result.rows[0] || null;
    }

    /**
     * Get task counts by status
     */
    static async getStatusCounts(agentId?: string): Promise<Record<TaskStatus, number>> {
        let query = 'SELECT status, COUNT(*)::int as count FROM tasks';
        const params: any[] = [];

        if (agentId) {
            query += ' WHERE agent_id = $1';
            params.push(agentId);
        }

        query += ' GROUP BY status';

        const result = await db.query(query, params);

        const counts: Record<string, number> = {
            pending: 0,
            processing: 0,
            completed: 0,
            failed: 0,
        };

        result.rows.forEach((row) => {
            counts[row.status] = row.count;
        });

        return counts as Record<TaskStatus, number>;
    }
}

File: backend/internal/models/task.go
package models

import (
	"database/sql"
	"encoding/json"
	"time"

	"go-backend/internal/database"

	"github.com/google/uuid"
)

type TaskStatus string

const (
	TaskStatusPending    TaskStatus = "pending"
	TaskStatusProcessing TaskStatus = "processing"
	TaskStatusCompleted  TaskStatus = "completed"
	TaskStatusFailed     TaskStatus = "failed"
)

type TaskPriority string

const (
	TaskPriorityHigh   TaskPriority = "high"
	TaskPriorityMedium TaskPriority = "medium"
	TaskPriorityLow    TaskPriority = "low"
)

type Task struct {
	ID          string          `json:"id"`
	AgentID     string          `json:"agent_id"`
	UserID      *string         `json:"user_id,omitempty"`
	Action      string          `json:"action"`
	Input       json.RawMessage `json:"input"`
	Output      json.RawMessage `json:"output,omitempty"`
	Status      TaskStatus      `json:"status"`
	Priority    TaskPriority    `json:"priority"`
	Error       *string         `json:"error,omitempty"`
	StartedAt   *time.Time      `json:"started_at,omitempty"`
	CompletedAt *time.Time      `json:"completed_at,omitempty"`
	CreatedAt   time.Time       `json:"created_at"`
}

type TaskRepository struct{}

func NewTaskRepository() *TaskRepository {
	return &TaskRepository{}
}

func (r *TaskRepository) Create(agentID, action string, input map[string]any, priority, userID string) (*Task, error) {
	id := uuid.New().String()
	inputJSON, err := json.Marshal(input)
	if err != nil {
		return nil, err
	}

	var userIDPtr *string
	if userID != "" {
		userIDPtr = &userID
	}

	query := `
		INSERT INTO tasks (id, agent_id, user_id, action, input, status, priority)
		VALUES ($1, $2, $3, $4, $5, $6, $7)
		RETURNING id, agent_id, user_id, action, input, status, priority, created_at
	`

	task := &Task{}
	err = database.DB.QueryRow(query, id, agentID, userIDPtr, action, inputJSON, TaskStatusPending, priority).Scan(
		&task.ID, &task.AgentID, &task.UserID, &task.Action, &task.Input, &task.Status, &task.Priority, &task.CreatedAt,
	)
	if err != nil {
		return nil, err
	}

	return task, nil
}

func (r *TaskRepository) FindByID(id string) (*Task, error) {
	query := `
		SELECT id, agent_id, user_id, action, input, output, status, priority, error, started_at, completed_at, created_at
		FROM tasks WHERE id = $1
	`

	task := &Task{}
	var output, errorMsg sql.NullString
	var startedAt, completedAt sql.NullTime

	err := database.DB.QueryRow(query, id).Scan(
		&task.ID, &task.AgentID, &task.UserID, &task.Action, &task.Input,
		&output, &task.Status, &task.Priority, &errorMsg, &startedAt, &completedAt, &task.CreatedAt,
	)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}

	if output.Valid {
		task.Output = json.RawMessage(output.String)
	}
	if errorMsg.Valid {
		task.Error = &errorMsg.String
	}
	if startedAt.Valid {
		task.StartedAt = &startedAt.Time
	}
	if completedAt.Valid {
		task.CompletedAt = &completedAt.Time
	}

	return task, nil
}

func (r *TaskRepository) FindAll(filters map[string]any) ([]Task, int, error) {
	query := `SELECT id, agent_id, user_id, action, input, output, status, priority, error, started_at, completed_at, created_at FROM tasks WHERE 1=1`
	countQuery := `SELECT COUNT(*)::int FROM tasks WHERE 1=1`
	args := []any{}
	argCount := 1

	if agentID, ok := filters["agentId"].(string); ok && agentID != "" {
		query += ` AND agent_id = $` + string(rune('0'+argCount))
		countQuery += ` AND agent_id = $` + string(rune('0'+argCount))
		args = append(args, agentID)
		argCount++
	}

	if status, ok := filters["status"].(string); ok && status != "" {
		query += ` AND status = $` + string(rune('0'+argCount))
		countQuery += ` AND status = $` + string(rune('0'+argCount))
		args = append(args, status)
		argCount++
	}

	query += ` ORDER BY created_at DESC`

	limit := 50
	offset := 0
	if l, ok := filters["limit"].(int); ok && l > 0 {
		limit = l
	}
	if o, ok := filters["offset"].(int); ok && o > 0 {
		offset = o
	}

	query += ` LIMIT $` + string(rune('0'+argCount))
	args = append(args, limit)
	argCount++

	query += ` OFFSET $` + string(rune('0'+argCount))
	args = append(args, offset)

	// Get total count
	var total int
	countArgs := args[:len(args)-2] // Exclude limit and offset
	err := database.DB.QueryRow(countQuery, countArgs...).Scan(&total)
	if err != nil {
		return nil, 0, err
	}

	// Get tasks
	rows, err := database.DB.Query(query, args...)
	if err != nil {
		return nil, 0, err
	}
	defer rows.Close()

	tasks := []Task{}
	for rows.Next() {
		task := Task{}
		var output, errorMsg sql.NullString
		var startedAt, completedAt sql.NullTime

		err := rows.Scan(
			&task.ID, &task.AgentID, &task.UserID, &task.Action, &task.Input,
			&output, &task.Status, &task.Priority, &errorMsg, &startedAt, &completedAt, &task.CreatedAt,
		)
		if err != nil {
			return nil, 0, err
		}

		if output.Valid {
			task.Output = json.RawMessage(output.String)
		}
		if errorMsg.Valid {
			task.Error = &errorMsg.String
		}
		if startedAt.Valid {
			task.StartedAt = &startedAt.Time
		}
		if completedAt.Valid {
			task.CompletedAt = &completedAt.Time
		}

		tasks = append(tasks, task)
	}

	return tasks, total, nil
}

func (r *TaskRepository) UpdateStatus(id string, status TaskStatus) error {
	query := `UPDATE tasks SET status = $1, started_at = NOW() WHERE id = $2`
	_, err := database.DB.Exec(query, status, id)
	return err
}

func (r *TaskRepository) UpdateCompleted(id string, output map[string]any) error {
	outputJSON, err := json.Marshal(output)
	if err != nil {
		return err
	}

	query := `UPDATE tasks SET status = $1, output = $2, completed_at = NOW() WHERE id = $3`
	_, err = database.DB.Exec(query, TaskStatusCompleted, outputJSON, id)
	return err
}

func (r *TaskRepository) UpdateFailed(id string, errorMsg string) error {
	query := `UPDATE tasks SET status = $1, error = $2, completed_at = NOW() WHERE id = $3`
	_, err := database.DB.Exec(query, TaskStatusFailed, errorMsg, id)
	return err
}

func (r *TaskRepository) GetStatusCounts(agentID string) (map[string]int, error) {
	query := `SELECT status, COUNT(*)::int FROM tasks`
	args := []any{}

	if agentID != "" {
		query += ` WHERE agent_id = $1`
		args = append(args, agentID)
	}

	query += ` GROUP BY status`

	rows, err := database.DB.Query(query, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	counts := map[string]int{
		"pending":    0,
		"processing": 0,
		"completed":  0,
		"failed":     0,
	}

	for rows.Next() {
		var status string
		var count int
		if err := rows.Scan(&status, &count); err != nil {
			return nil, err
		}
		counts[status] = count
	}

	return counts, nil
}
